### PR TITLE
Fix `NodeGroup` taints updates 

### PR DESCRIPTION
**Pull request by speedfl**
Issue: https://github.com/aws-controllers-k8s/community/issues/1950

Description of changes:

Add custom diff for taints + tests

### CODE REVIEW

File: pkg/resource/nodegroup/delta.go [hunk 1]
```diff
@@ -189,9 +189,29 @@ func newResourceDelta(
...
+					}
+				}
+				if !matched {
+					delta.Add("Spec.Taints", a.ko.Spec.Taints, b.ko.Spec.Taints)
+					break
+				}
+			} 
+		}
 	}
+
```
Comment (a-hilaly) [inline pkg/resource/nodegroup/delta.go:214]: This code is generated, so next time this controller is regenerated we will lose all this code :). Ideally you need to instruct the code generator on how to inject the new logic into `delta.go`
1. First we need to drop the current implementation: To do that you'll have add `compare: false` into the `NodeGroup` resource:
```yaml
  Nodegroup:
    fields:
      compare: false
```

2. We need to ask the code-generator to inject a new comparison function that will use your implementation to find the delta (which is already done - so no need to worry about this)
3. Add your logic to https://github.com/aws-controllers-k8s/eks-controller/blob/main/pkg/resource/nodegroup/hook.go#L80 (note that this function is already doing some defaulting)

Make sure to hit `make build-controller SERVICE=eks` from the code-generator repository everytime you modify the `generator.yaml`

File: pkg/resource/nodegroup/delta_test.go
Comment (a-hilaly) [inline pkg/resource/nodegroup/delta_test.go:1]: Neat!

**Comment by speedfl** [comment]
@a-hilaly I see that e2e failed. But issue is related to subnets. Is it a  flacky test ?

**Comment by a-hilaly** [comment]
Hi @speedfl - i just fixed the issue with NodeGroup failing tests, after a rebase we'll have more clarity on what's needed :)

**Comment by a-hilaly** [comment]
>@a-hilaly I see that e2e failed. But issue is related to subnets. Is it a flacky test ?
The subnet cleanup is faulty indeed but still if tests pass, it should be fine

File: helm/Chart.yaml [hunk 2]
```diff
@@ -1,8 +1,8 @@
 apiVersion: v1
 name: eks-chart
 description: A Helm chart for the ACK service controller for Amazon Elastic Kubernetes Service (EKS)
-version: 1.0.5
-appVersion: 1.0.5
+version: 0.0.0-non-release-version
+appVersion: 0.0.0-non-release-version
```
Comment (a-hilaly) [inline helm/Chart.yaml:5]: A `git fetch upstream main --tags` should will set the right tags here

File: pkg/resource/nodegroup/hook.go
Comment (a-hilaly) [inline pkg/resource/nodegroup/hook.go:1]: Great stuff!

File: pkg/resource/nodegroup/hook_test.go
Comment (a-hilaly) [inline pkg/resource/nodegroup/hook_test.go:1]: ❤️

File: pkg/resource/nodegroup/hook_test.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+					},
+				},
+			},
+		},
+	}
+
+	delta = &ackcompare.Delta{}
+	customPreCompare(delta, a, b)
+	assert.True(t, delta.DifferentAt("Spec.Taints"), "Taints have different length")
+}
```
Comment (a-hilaly) [inline pkg/resource/nodegroup/hook_test.go:134]: Can we also add some end2end tests, for taints updates?

File: pkg/resource/nodegroup/hook.go [hunk 4]
```diff
@@ -77,6 +79,27 @@ func customPreCompare(
...
+					if reflect.DeepEqual(taintA, taintB) {
+						matched = true
+						break
+					}
+				}
+				if !matched {
+					delta.Add("Spec.Taints", a.ko.Spec.Taints, b.ko.Spec.Taints)
+					break
+				}
+			}
```
Comment (a-hilaly) [inline pkg/resource/nodegroup/hook.go:100]: how about the taints that are in B, but not in A?

Comment (speedfl) [reply pkg/resource/nodegroup/hook.go:100]: Just above I am checking that length are equals.

File: pkg/resource/nodegroup/hook_test.go (hunk 3 above)
Comment (speedfl) [reply pkg/resource/nodegroup/hook_test.go:134]: I will check this!

Comment (speedfl) [reply pkg/resource/nodegroup/hook_test.go:134]: Done

File: helm/Chart.yaml (hunk 2 above)
Comment (speedfl) [reply helm/Chart.yaml]: Done

**Review by a-hilaly — APPROVED** [review]
Great work on this! Thank you @speedfl ~
/approve

**Comment by a-hilaly** [comment]
@speedfl Can you please send a release PR? `RELEASE_VERSION=v1.1.1 ./scripts/build-controller-release.sh eks` should do the job :)

**Comment by speedfl** [comment]
@a-hilaly the script hangs... Could you please run it ?

**Pull request merged**